Mathlib/Algebra/Module/Lattice.lean

```diff
@@ -0,0 +1,198 @@
+    exact Submodule.smul_mem_pointwise_smul _ _ _ (by trivial)
+
+end
+
+section Field
+
+variable {K : Type*} [Field K] [Algebra R K]
+
+lemma _root_.Submodule.span_range_eq_top_of_injective_of_rank {M N : Type u} [IsDomain R]
+    [IsFractionRing R K] [AddCommGroup M] [Module R M]
+    [AddCommGroup N] [Module R N] [Module K N] [IsScalarTower R K N] [Module.Finite K N]
+    {f : M →ₗ[R] N} (hf : Function.Injective f)
+    (h : Module.rank R M = Module.rank K N) :
+    Submodule.span K (LinearMap.range f : Set N) = ⊤ := by
+  obtain ⟨s, hs, hli⟩ := exists_set_linearIndependent R M
+  replace hli := hli.map' f (LinearMap.ker_eq_bot.mpr hf)
+  rw [LinearIndependent.iff_fractionRing (R := R) (K := K)] at hli
+  rw [h, ← Module.finrank_eq_rank, Cardinal.mk_eq_nat_iff_fintype] at hs
+  obtain ⟨hfin, hcard⟩ := hs
+  have hsubset : Set.range (fun x : s ↦ f x.val) ⊆ (LinearMap.range f : Set N) := by
+    rintro x ⟨a, rfl⟩
+    simp
+  have hcard : Fintype.card ↑s = Module.finrank K N := by simpa
+  rw [eq_top_iff, ← LinearIndependent.span_eq_top_of_card_eq_finrank' hli hcard]
+  exact Submodule.span_mono hsubset
+
+variable (K) {V : Type*} [AddCommGroup V] [Module K V] [Module R V] [IsScalarTower R K V]
+
+/-- Any basis of an `R`-lattice in `V` defines a `K`-basis of `V`. -/
+noncomputable def _root_.Basis.extendOfIsLattice [IsFractionRing R K] {κ : Type*}
+    {M : Submodule R V} [IsLattice K M] (b : Basis κ R M) :
+    Basis κ K V :=
+  have hli : LinearIndependent K (fun i ↦ (b i).val) := by
+    rw [← LinearIndependent.iff_fractionRing (R := R), linearIndependent_iff']
+    intro s g hs
+    simp_rw [← Submodule.coe_smul_of_tower, ← Submodule.coe_sum, Submodule.coe_eq_zero] at hs
+    exact linearIndependent_iff'.mp b.linearIndependent s g hs
+  have hsp : ⊤ ≤ span K (Set.range fun i ↦ (M.subtype ∘ b) i) := by
+    rw [← Submodule.span_span_of_tower R, Set.range_comp, ← Submodule.map_span]
+    simp [b.span_eq, Submodule.map_top, span_eq_top]
+  Basis.mk hli hsp
+
+@[simp]
+lemma _root_.Basis.extendOfIsLattice_apply [IsFractionRing R K] {κ : Type*}
+    {M : Submodule R V} [IsLattice K M] (b : Basis κ R M) (k : κ) :
+    b.extendOfIsLattice K k = (b k).val := by
+  simp [Basis.extendOfIsLattice]
+
+variable [IsDomain R]
+
+/-- A finitely-generated `R`-submodule of `V` that has rank the `K`-rank of `V` is
+a lattice. -/
+lemma of_rank [Module.Finite K V] [IsFractionRing R K] {M : Submodule R V}
+    (hfg : M.FG) (hr : Module.rank R M = Module.rank K V) : IsLattice K M where
+  fg := hfg
+  span_eq_top := by
+    simpa using Submodule.span_range_eq_top_of_injective_of_rank M.injective_subtype hr
+
+/-- The supremum of two lattices is a lattice. -/
+instance sup (M N : Submodule R V) [IsLattice K M] [IsLattice K N] :
+    IsLattice K (M ⊔ N) where
+  fg := Submodule.FG.sup IsLattice.fg IsLattice.fg
+  span_eq_top := by
+    rw [eq_top_iff]
+    exact le_trans (b := span K M) (by rw [IsLattice.span_eq_top]) (Submodule.span_mono (by simp))
```

Comment:
Maybe you can prove a more general result first: a finitely generated `R`-module which contains a lattice is a lattice. And this is just a corollary of it. To prove this, maybe you can generalize `of_rank` (as well as `span_range_eq_top_of_injective_of_rank`) to `of_rank_ge` which requires `Module.rank R M ≥ Module.rank K V`.